core/src/storage/cache.py

```diff
@@ -1,36 +1,68 @@
 from trezor.crypto import random
 
 if False:
-    from typing import Optional
+    from typing import Optional, Dict, List
+
+_MAX_SESSIONS_COUNT = 10
 
 APP_COMMON_SEED = 0
 APP_COMMON_SEED_WITHOUT_PASSPHRASE = 1
 APP_CARDANO_ROOT = 2
 APP_MONERO_LIVE_REFRESH = 3
 
-_cache_session_id = None  # type: Optional[bytes]
-_cache = {}
+_active_session_id = None  # type: Optional[bytes]
+_caches = {}  # type: Dict[bytes, Dict[int, Any]]
+_session_ids = []  # type: List[bytes]
 
 if False:
     from typing import Any
 
 
-def get_session_id() -> bytes:
-    global _cache_session_id
-    if not _cache_session_id:
-        _cache_session_id = random.bytes(32)
-    return _cache_session_id
+def _move_session_ids_queue(session_id: bytes) -> None:
+    # Move the LRU session ids queue.
+    if session_id in _session_ids:
+        _session_ids.remove(session_id)
+    elif len(_session_ids) == _MAX_SESSIONS_COUNT:
+        remove_session_id = _session_ids.pop()
+        del _caches[remove_session_id]
+
+    _session_ids.insert(0, session_id)
+
+
+def start_session(received_session_id: bytes = None) -> bytes:
+    if received_session_id and received_session_id in _session_ids:
+        session_id = bytes(received_session_id)
```

Comment:
why the `bytes()`?

Reply:
Because it is used as a key to to the dict. And dict's key needs to be immutable, which `bytearray` is not. In other words, `__hash__` method is missing. Am I mistaken?

Reply:
But if you are asking why am I receiving `bytearray` instead of `bytes` in the Initialize message, I'm not sure. Anyway I'll add it to the `handle_Initialize` function so the type hint is correct.

Reply:
yes, the second one :) converting this in handle_Initialize would be preferable, as the bytearrays are coming in from protobuf, but this code doesn't know that